src/cascadia/TerminalSettingsModel/ApplicationState.h

```diff
@@ -0,0 +1,45 @@
+/*++
+Copyright (c) Microsoft Corporation
+Licensed under the MIT license.
+
+Module Name:
+- ApplicationState.h
+
+--*/
+#pragma once
+
+#include "ApplicationState.g.h"
+#include "../inc/cppwinrt_utils.h"
+
+namespace winrt::Microsoft::Terminal::Settings::Model::implementation
+{
+    struct ApplicationState : ApplicationStateT<ApplicationState>
+    {
+        ApplicationState(std::filesystem::path path) :
+            _path{ path } {}
+
+        static Microsoft::Terminal::Settings::Model::ApplicationState GetForCurrentApp();
+        void Reload();
+        void Reset();
+
+        void Commit();
+
+        GETSET_PROPERTY(bool, CloseAllTabsWarningDismissed, false);

```

Comment:
Consider splitting between the "state storage" and the actual state.
BTW, probably, the storage logic doesn't belong to the model, as we always want to keep model as thin as possible.

Reply:
Now is when I should admit that, despite more than a decade in software, _I'm bad at software architecture_. :smile:

Does it seem reasonable to have `static ApplicationStateStorage::Current()` and `ApplicationStateStorage::ApplicationState()`, which returns one of these? xxxStorage handles the file/path resolution and loading, xxxState maintains its serialization/deserialization logic?

I'm spitballing hear, so bear with me.

```
runtimeclass ApplicationStateStorage {
    // ApplicationStateStorage(String path); // maybe?
    static ApplicationStateStorage GetForCurrentApplication();
    ApplicationState LoadState(); // should be async, but I don't want to crack that can of worms right now[1]
    IAsyncAction WriteStateAsync(ApplicationState s);
    IAsyncAction DeleteStateAsync(); // erases state from disk; consumers should discard their held state?
}
```

The consumer would have to hold on to the state this way. Not sure I like that, since it's supposed to be this "always-available" fixture.

Our whole model is pretty lousy with mixed concepts right now. :(

1: the whole of settings load is synchronous and blocks startup (😬), and this would unfortunately be implicated in that. that's a disentanglement for another month.

Reply:
I am bad in coding so I will write English :smile:
1. Introduce `FileStorage<T> : IStorage<T>` - that receives file path and `ISerializer<T>` upon construction and holds the snapshot of type `T`
2. The aforementioned `ISerializer<T>` - is an interface for serializing / deserlializing `T`
3. Introduce `ApplicationState` and `ApplicationStateSerializer  : ISerializer<ApplicationState>`
4. Store a static instance of type `FileStorage<ApplicationState>` if you need a singleton

In the description above only `ApplicationState` is in the model. WDYT?